### DISCUSSION TITLE
(Lack of) efficient multi-threading on the CPU and `vmap` + `jit`

### DISCUSSION THREAD

**Discussion by toddpocuca** [jax-ml/jax#34267] [Q&A]
I am writing a library for Bayesian inference with JAX and I have come across an issue where I do not think `vmap` and `jit` will efficiently use the CPU.

For context, I am using a variational inference method for approximating a posterior distribution. In order to do this I must compute the gradient of the ELBO to iteratively optimize the approximation. However, the ELBO cannot be computed analytically for many models and thus a noisy (but unbiased) estimate must be computed instead. Here is a snippet of my code for doing this:
```py
    @eqx.filter_jit
    def elbo_grad(self, n: int, batch_size: int, key: PRNGKeyArray) -> Self:
        dyn, static = eqx.partition(self, self.filter_spec)

        # Define ELBO function
        def elbo(dyn: Self, n: int, key: PRNGKeyArray) -> Scalar:
            self = eqx.combine(dyn, static)

            # Split key
            keys = jr.split(key, n // batch_size)

            # Split ELBO calculation into batches
            def batched_elbo(carry: None, batch_key: PRNGKeyArray) -> tuple[None, Array]:
                # Draw from variational distribution
                draws: Array = self.base.sample(batch_size, sir = False, key = batch_key)

                # Evaluate posterior and variational densities
                batched_post_evals, batched_vari_evals = self.__eval(draws)

                # Compute batched ELBO evals
                batched_elbo_evals: Array = batched_post_evals - batched_vari_evals

                return None, batched_elbo_evals

            # Compute ELBO evals
            elbo_evals = lax.scan(
                batched_elbo,
                init=None,
                xs=keys,
                length=n // batch_size
            )[1].flatten()

            # Compute average of ELBO estimates
            return jnp.mean(elbo_evals)

        # Map to its gradient
        elbo_grad: Callable[
            [Self, int, PRNGKeyArray], Self
        ] = eqx.filter_grad(elbo)

        return elbo_grad(dyn, n, key)
```

I have noticed that JAX will effectively parallelize the posterior density evaluation (which makes sense as it usually just involves the same or similar operations applied to a collection of values all independently). However, for some models a large number of gradient samples (`n`) are needed in order to get a stable estimator of the ELBO, and thus it would be useful to parallelize over this outer level of work as well (especially when the posterior density evaluation is small enough that it is inefficient to parallelize at that level of work). 

So the `__eval` method allows batches of draws to be processed together usually using `vmap`. But even when I set the number of gradient samples to be high (and the `batch_size` to be equal to `n`, so the `scan` is effectively doing nothing), I do not see my cores being used more efficiently, instead I just observe each P-core (I am on apple silicon) with 50% usage.

What can I do to make this more efficient? I do see my GPU (separate device with NVIDIA GPU for testing) having higher usage with increasing gradient samples but I do not observe the same on my CPU.
